### ISSUE TITLE
[POC] arithmetic operations over GPU

### ISSUE THREAD

**Issue by GMartigny** [gpujs/gpu.js#427]
Hi,

I liked the idea of running JS on GPU and wanted take a stab at it.
In order to start small and learn your API, I write a couple of kernels to execute simple arithmetic operations (max and sum).
https://github.com/GMartigny/gpu.maths
But, when I benchmark those against native node operation, I get poor results.

![image](https://user-images.githubusercontent.com/2543511/53484333-e47c7800-3a83-11e9-9036-81265044d7f8.png)

What's the problem here ?
Is this use of gpu.js not right ?
Does my GPU less performant than my CPU (not impossible) ?
Am I doing something wrong with the API ?

**Comment by hmmhmmhm** [comment]
I'm curious, too

**Comment by robertleeplummerjr** [comment]
Try invoking build before benchmarking `kernel.build(arg1)`, and then running it as a comparison.  It may be a bit faster in that case, and too it can be an option when using textures (for example, if I've done a bunch of operations already on the GPU, and need to calculate sum or max, without transferring the values back to the CPU), but likely this small operation alone wouldn't warrant using the GPU, because it returns to a single output, which means it isn't going to run in tandem, but rather in a single thread. 

In short: the difference would be the processing speed of 1 CPU thread vs 1 GPU thread.

**Comment by GMartigny** [comment]
For the records, using the `kernel.build()` operation gains a few ms, but nothing huge (about 10ms => 7%).
I added two matrix operation (addition and multiplication) to leverage the speed of parallelization. Results on my computer are better, but still under a naive nested loopdy-loop solution.

![image](https://user-images.githubusercontent.com/2543511/53502162-cfb4da00-3aad-11e9-8ac3-6ec59de7e523.png)

I'll try on a better GPU later.

**Comment by robertleeplummerjr** [comment]
When adding the new operations, do they still go to one output as before?

**Comment by GMartigny** [comment]
I mean, each operation are not piped into each others. I just checked 4 different kernel running time against their naive counterpart.

I had the time to check on another computer. Results are expected and strange at the same time. We can see no improvement on `Array => Number` operations (as expected because of mono-threading). There's however, a huge gain on `Matrix => Matrix` operation, but only for multiplication (about 5x quicker). Matrix addition stayed on the same foot.

![image](https://user-images.githubusercontent.com/2543511/53506466-4fdf3d80-3ab6-11e9-88c1-e653fd82bb1f.png)

My hypothesis is that, for trivial operation, minimal complexity should be strictly above `O(n²)`.

**Comment by robertleeplummerjr** [comment]
>My hypothesis is that, for trivial operation, minimal complexity should be strictly above `O(n²)`.

I think your assumption is 100% correct, because a video card, if forced to run in a single thread, can only achieve its highest clock.  Example: https://www.nvidia.com/en-us/geforce/graphics-cards/rtx-2060/

`Boost Clock | 1680 MHz`

You'd be forcing a very powerful card to do very little of anything.  Closing as well answered previously.
